models/transformer/decoder_impl.hpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * @file models/transformer/decoder_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of the Transformer Decoder class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_TRANSFORMER_DECODER_IMPL_HPP
+#define MODELS_TRANSFORMER_DECODER_IMPL_HPP
+
+#include "decoder.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType,
+    class ActivationFunction, typename RegularizerType>
+TransformerDecoder<InputDataType, OutputDataType, ActivationFunction,
+RegularizerType>::TransformerDecoder() :
+    dModel(0),
+    numHeads(0),
+    dimFFN(0),
+    dropout(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType,
+    class ActivationFunction, typename RegularizerType>
+TransformerDecoder<InputDataType, OutputDataType, ActivationFunction,
+RegularizerType>::TransformerDecoder(const size_t dModel,
+    const size_t numHeads,
+    const size_t dimFFN,
+    const ElemType dropout) :
+    dModel(dModel),
+    numHeads(numHeads),
+    dimFFN(dimFFN),
+    dropout(dropout)
+{
+  attnBlock1.Add<MultiheadAttention<InputDataType, OutputDataType, 0, RegularizerType> >(dModel,
+      numHeads, dimFFN, dropout, false);
+  attnBlock1.Add<Dropout<InputDataType, OutputDataType> >(dropout);
+
+  attnBlock2.Add<MultiheadAttention<InputDataType, OutputDataType, 1, RegularizerType> >(dModel,
+      numHeads, dimFFN, dropout, false);
+  attnBlock2.Add<Dropout<InputDataType, OutputDataType> >(dropout);
+
+  ffnBlock.Add<Linear<InputDataType, OutputDataType, RegularizerType> >(dModel, dimFFN);
+  ffnBlock.Add<ActivationFunction>();
+  ffnBlock.Add<Linear<InputDataType, OutputDataType, RegularizerType> >(dimFFN, dModel);
+  ffnBlock.Add<Dropout<InputDataType, OutputDataType> >(dropout);
+}
+
+template <typename InputDataType, typename OutputDataType,
+    class ActivationFunction, typename RegularizerType>
+void TransformerDecoder<InputDataType, OutputDataType, ActivationFunction,
+RegularizerType>::Assemble()
```

Comment:
Most probably I missed something, so I guess there is a good reason to not put this into the constructor?

Reply:
Ahh, yes. Because the `Concatenate` layer requires the key matrix (line 64) to be concatenated (line 67 & 68) and how can one assign `key` if object is not yet constructed? (`key` is the output of the Encoder). Though it was not required for `Encoder`, yet I made them separate to maintain the same API. The name of the method might not be a good choice 😁, so I am open to any other name for this method. When I named it, in my mind - "Avengers! Assemble" :smiley:

Reply:
Ahh, thanks for the clarification, I like the name.